metrics.go

```diff
@@ -0,0 +1,191 @@
+type rateLimitMetrics struct {
+	declinedTotal *prometheus.CounterVec
+	requestsTotal *prometheus.CounterVec
+	processTime   *prometheus.HistogramVec
+	keysTotal     *prometheus.GaugeVec
+	config        *prometheus.CounterVec
+}
+
+var (
+	// Metrics registration sync to ensure we only register once
+	metricsOnce sync.Once
+	// Global metrics instance
+	globalMetrics *rateLimitMetrics
+)
+
+// initializeMetrics creates and registers all rate limit metrics with Caddy's internal registry
+func initializeMetrics(registry prometheus.Registerer) *rateLimitMetrics {
+	const ns, sub = "caddy", "rate_limit"
+
+	factory := promauto.With(registry)
+
+	return &rateLimitMetrics{
+		// rate_limit_declined_requests_total - Total number of requests declined with HTTP 429
+		declinedTotal: factory.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: ns,
+				Subsystem: sub,
+				Name:      "declined_requests_total",
+				Help:      "Total number of requests for which rate limit was applied (Declined with HTTP 429 status code returned).",
+			},
+			[]string{"zone", "key"},
+		),
+
+		// rate_limit_requests_total - Total number of requests that passed through the Rate Limit module
+		requestsTotal: factory.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: ns,
+				Subsystem: sub,
+				Name:      "requests_total",
+				Help:      "Total number of requests that passed through Rate Limit module (both declined & processed).",
+			},
+			[]string{"zone", "key"},
+		),
+
+		// rate_limit_process_time_seconds - Time taken to process rate limiting for each request
+		processTime: factory.NewHistogramVec(
+			prometheus.HistogramOpts{
+				Namespace: ns,
+				Subsystem: sub,
+				Name:      "process_time_seconds",
+				Help:      "A time taken to process rate limiting for each request.",
+				Buckets:   []float64{.001, .005, .01, .025, .05, .1, .25, .5, 1},
+			},
+			[]string{"zone", "key"},
+		),
+
+		// rate_limit_keys_total - Total number of keys that each RL zone contains
+		keysTotal: factory.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Namespace: ns,
+				Subsystem: sub,
+				Name:      "keys_total",
+				Help:      "Total number of keys that each RL zone contains. (This metric is collected in the background for each zone.)",
+			},
+			[]string{"zone"},
+		),
+
+		// rate_limit_config - Shows configuration of the rate limiter module
+		config: factory.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: ns,
+				Subsystem: sub,
+				Name:      "config",
+				Help:      "Shows configuration of the rate limiter module. Reported only once on bootstrap as configuration is not dynamically configurable.",
+			},
+			[]string{"zone", "max_events", "window"},
+		),
+	}
+}
+
+// registerMetrics registers all rate limit metrics with the provided Prometheus registry
+func registerMetrics(reg prometheus.Registerer) error {
+	var err error
+	metricsOnce.Do(func() {
+		globalMetrics = initializeMetrics(reg)
+	})
```

Comment:
This will stop working on config reload, and no metrics will be registered or emitted. Unfortunately, the only way to manage it for now is to check for duplicate registration error, and ignore it.